### PR TITLE
Remove sql injection vulnerability 

### PR DESCRIPTION
**Pull request by JamesOsborn-SE**

### CODE REVIEW

**Comment by zefr0x** [comment]
Thanks for your contribution and thanks for giving me this information. 😄

I ran into this issue but didn't find any solution, then I left it, because I didn't see it as dangerous since the database used here is just a copy and the runner's input should be trusted. However, thanks agian.

File: bookmarks.py [hunk 1]
```diff
@@ -55,30 +55,31 @@ def connect_to_database(self) -> None:
...
+
         else:
-            # VULN sql injection.
+            query += " WHERE A.title LIKE ? "
             query += (
-                ' ORDER BY instr(LOWER(A.title), LOWER("%s")) ASC, visit_count DESC'
-                % term
+                "ORDER BY instr(LOWER(A.title), LOWER(?)) ASC, visit_count DESC"
             )
+            args = ('%'+term+'%', term)
```
Comment (zer0-x) [inline bookmarks.py:75]: I want to ask why are the two `%` important here?

File: bookmarks.py [hunk 2]
```diff
@@ -55,30 +55,31 @@ def connect_to_database(self) -> None:
             self.cursor = self.con.cursor()
         return None
 
-    def search(self, term) -> list:
+    def search(self, term = "") -> list:
         """Search for a term in the database and return the results."""
-        # VULN sql injection.
+        args = (term)
```
Comment (zefr0x) [inline bookmarks.py]: Is this line important? If it was, I think it should be just an empty tuple, but what I can see is that its a sitring.
If you wanted a tuple should it be `(term,)` ?

I think it should be just `()`, which is just an empty tuple.

Comment (JamesOsborn-SE) [reply bookmarks.py]: parentheses aren't needed but the assignment is still needed so args isn't `referenced before assignment`
technically you could reassign the terms but it makes for confusing code.

File: bookmarks.py [hunk 3]
```diff
@@ -55,30 +55,31 @@ def connect_to_database(self) -> None:
...
         """Search for a term in the database and return the results."""
-        # VULN sql injection.
         query = (
             "SELECT A.title, B.url FROM moz_bookmarks AS A"
             + " JOIN moz_places AS B ON(A.fk = B.id)"
-            + ' WHERE A.title LIKE "%%%s%%"'
-        ) % term
+        )
+
+        args = term
```
Comment (zefr0x) [inline bookmarks.py]: But what's the point of `args = term` if we don't need the args? It's just for avoiding the error.

When `term == ""` we don't need any args so it should be an empty tuple. Why should we pass `term` to `self.cursor.execute` when we don't have any `?` in the sql query?

It should be:
```python
args = ()
```
rather then:
```python
args = term
```

Am I wrong?

Comment (JamesOsborn-SE) [reply bookmarks.py]: that's a good point

Comment (JamesOsborn-SE) [reply bookmarks.py]: I pushed up a change I'll let you resolve this conversation if you're happy with the changes.

**Pull request merged**